ws_validator.js

```diff
@@ -34,314 +34,346 @@ var ws;
+                  err: err,
+                  profile: err.profile
+                });
+              }
 
-        var log_prepend = 'user ' + payload.username + ':';
-        var log = console.log.bind(console, log_prepend.blue);
+              log("Change Password attempt failed. Reason: " + "unexpected error".red);
 
-        log('Attempting change_password.');
+              if (err.inner && err.inner.stack) {
+                console.error('Inner error:', err.inner.stack);
+              } else {
+                console.log(err);
+              };
 
-        users.changePassword(payload.username, payload.password, function (err, profile) {
-          if (err) {
-            if (err instanceof InsufficientAccessRightsError || err instanceof PasswordComplexityError) {
-              log("Change Password attempt failed. Reason: " + err.message.red);
-              return ws.sendEvent(payload.pid + '_change_password_result', {
-                err: err,
-                profile: err.profile
+              ws.sendEvent(payload.pid + '_change_password_result', {
+                err: err
               });
+              return exit(1);
             }
 
-            log("Change Password attempt failed. Reason: " + "unexpected error".red);
-
-            if (err.inner && err.inner.stack) {
-              console.error('Inner error:', err.inner.stack);
-            } else {
-              console.log(err);
-            };
-
+            log('Password Change succeeded.');
             ws.sendEvent(payload.pid + '_change_password_result', {
-              err: err
+              profile: profile
             });
-            return exit(1);
-          }
-
-          log('Password Change succeeded.');
-          ws.sendEvent(payload.pid + '_change_password_result', {
-            profile: profile
           });
         });
       });
-    });
-  }
+    }
 
-  function authenticate_connector() {
-    var token = jwt.sign({}, cert.key, {
-      algorithm: 'RS256',
-      expiresIn: '1m',
-      issuer: nconf.get('CONNECTION'),
-      audience: nconf.get('REALM')
-    });
+    function authenticate_connector() {
+      var token = jwt.sign({}, cert.key, {
+        algorithm: 'RS256',
+        expiresIn: '1m',
+        issuer: nconf.get('CONNECTION'),
+        audience: nconf.get('REALM')
+      });
 
-    ws.sendEvent('authenticate', {
-      jwt: token
-    });
-  };
+      ws.sendEvent('authenticate', {
+        jwt: token
+      });
+    };
+  })
+}
+
+async function reconnect() {
+  try {
+    console.log('Connecting to websocket...');
+    await setupWebsocket();
+    console.log('Connected to websocket');
+  } catch (err) {
+    console.error('websocket connection error:', new Error(err).message)
+  }
 }
 
-setupWebsocket();
+reconnect();
+
+// check periodically if reconnection is needed
+setInterval(() => {
+  if (!openedSocket) {
+    console.log('websocket deconnected. reconnecting')
```

Comment:
Sorry, it's a bit hard to read with github freaking out with the diff but when we have this `openedSocket == false` condition, are we closing up any dead resources for the socket / client? I am not sure how much cleanup is needed with the lib we are using but yeah.

Reply:
this will happen whenever a socket is in error. I removed all listeners from the dead socket, but I'll see if I can perform a close() on it just before assigning null and replacing it with a new

Reply:
added a call to terminate() based on `ws` docs